aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -327,6 +328,42 @@ Tensor& prod_out(Tensor& result, const Tensor& self, Dimname dim,
   return at::prod_out(result, self, dimname_to_position(self, dim), keepdim, opt_dtype);
 }
 
+static Tensor& nanprod_out_impl(Tensor& result, const Tensor& self, IntArrayRef dim,
+                        bool keepdim, c10::optional<ScalarType> opt_dtype) {
+  ScalarType dtype = get_dtype(result, self, opt_dtype, true);
+  auto iter = make_reduction("nanprod", result, self, dim, keepdim, dtype);
+  if (iter.numel() == 0) {
+    result.fill_(1);
+  } else {
+    nanprod_stub(iter.device_type(), iter);
+  }
+  return result;
+}
+
+Tensor nanprod(const Tensor& self, int64_t dim, bool keepdim, c10::optional<ScalarType> dtype) {
+  Tensor result;
+  native::nanprod_out_impl(result, self, dim, keepdim, dtype);
+  return result;
+}
+
+Tensor nanprod(const Tensor& self, c10::optional<ScalarType> dtype) {
+  Tensor result;
+  return at::native::nanprod_out_impl(result, self, {}, false, dtype);
+}
+
+Tensor& nanprod_out(Tensor& result, const Tensor& self, int64_t dim, bool keepdim, c10::optional<ScalarType> dtype) {
+  return at::native::nanprod_out_impl(result, self, dim, keepdim, dtype);
+}
+
+Tensor nanprod(const Tensor& self, Dimname dim, bool keepdim, c10::optional<ScalarType> dtype) {
+  return at::nanprod(self, dimname_to_position(self, dim), keepdim, dtype);
+}
+
+Tensor& nanprod_out(Tensor& result, const Tensor& self, Dimname dim,
```

Comment:
You don't need to support Named Tensors in this first PR, but if you want to you'll need to add at least one named tensor-specific test later.

Reply:
I have removed the support of named tensors in this PR. :)